src/Orleans.Transactions/InClusterTM/TransactionManager.cs

```diff
@@ -0,0 +1,695 @@
+                        transactions.Add(tx);
+                    }
+
+                    processed += currentBatchSize;
+
+                    // If the transaction involved writes, send a commit notification to each resource which performed
+                    // a write and wait for acknowledgement.
+                    if (resources.Count > 0)
+                    {
+                        // Send commit notifications to all of the resources involved.
+                        var completion = new MultiCompletionSource(resources.Count);
+                        foreach (var resource in resources)
+                        {
+                            NotifyResourceOfCommit(resource.Key, resource.Value, completion);
+                        }
+
+                        // Wait for the commit notifications to be acknowledged by the resources.
+                        await completion.Task;
+                    }
+
+                    // Mark the transactions as checkpointed.
+                    foreach (var tx in transactions)
+                    {
+                        lock (tx)
+                        {
+                            tx.State = TransactionState.Checkpointed;
+                            tx.HighestActiveTransactionIdAtCheckpoint = activeTransactionsTracker.GetHighestActiveTransactionId();
+                        }
+                    }
+
+                    // Allow the transaction log to be truncated.
+                    this.checkpointedLSN = lsn;
+                }
+            }
+            catch (Exception e)
+            {
+                // Retry all failed checkpoint operations.
+                foreach (var tx in transactions) this.checkpointRetryQueue.Enqueue(tx);
+
+                this.Logger.Error(0, "Failure during checkpoint", e);
+                throw;
+            }
+
+            return total > 0;
+        }
+
+        private static void NotifyResourceOfCommit(ITransactionalResource resource, long transaction, MultiCompletionSource completionSource)
+        {
+            resource.Commit(transaction)
+                    .ContinueWith(
+                        (result, state) =>
+                        {
+                            var completion = (MultiCompletionSource)state;
+                            if (result.Exception != null) completion.SetException(result.Exception);
+                            else completion.SetOneResult();
+                        },
+                        completionSource,
+                        CancellationToken.None,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
+        }
+
+        private void GC(object args)
+        {
+            //
+            // Truncate log
+            //
+            if (checkpointedLSN > 0)
+            {
+                try
+                {
+                    transactionLog.TruncateLog(checkpointedLSN - 1).Wait();
```

Comment:
It's not 'good' that we're blocking a threadpool thread here. Preferably we would use an async timer and await this.